test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java

```diff
@@ -0,0 +1,203 @@
+
+    // The size of one CEN header, including the name and the extra field
+    static final int CEN_HEADER_SIZE = ZipFile.CENHDR + NAME_LENGTH + MAX_EXTRA_FIELD_SIZE;
+
+    // The number of entries needed to exceed the MAX_CEN_SIZE
+    static final int NUM_ENTRIES = (MAX_CEN_SIZE / CEN_HEADER_SIZE) + 1;
+
+    // Helps SparseOutputStream detect write of the last CEN entry
+    private static final String LAST_COMMENT = "LastCEN";
+    private static final byte[] LAST_COMMENT_BYTES = LAST_COMMENT.getBytes(StandardCharsets.UTF_8);
+
+    // Zip file to create for testing
+    private File hugeZipFile;
+
+    /**
+     * Create a zip file with a CEN size which does not fit within a Java byte array
+     */
+    @BeforeTest
+    public void setup() throws IOException {
+        hugeZipFile = new File("cen-too-large.zip");
+        hugeZipFile.deleteOnExit();
+
+        try (OutputStream out = new SparseOutputStream(new FileOutputStream(hugeZipFile));
+             ZipOutputStream zip = new ZipOutputStream(out)) {
+
+            // Keep track of entries so we can update extra data before the CEN is written
+            ZipEntry[] entries = new ZipEntry[NUM_ENTRIES];
+
+            // Add entries until MAX_CEN_SIZE is reached
+            for (int i = 0; i < NUM_ENTRIES; i++) {
+                // Create a fixed-length name for the entry
+                String name = Integer.toString(i);
+                name = "0".repeat(NAME_LENGTH - name.length()) + name;
+
+                // Create and track the entry
+                ZipEntry entry = entries[i] = new ZipEntry(name);
+
+                // Use STORED for faster processing
+                entry.setMethod(ZipEntry.STORED);
+                entry.setSize(0);
+                entry.setCrc(0);
+
+                // Set the time/date field for faster processing
+                entry.setTimeLocal(TIME_LOCAL);
+
+                if (i == NUM_ENTRIES -1) {
+                    // Help SparseOutputStream detect the last CEN entry write
+                    entry.setComment(LAST_COMMENT);
+                }
+                // Add the entry
+                zip.putNextEntry(entry);
+
+
+            }
+            // Finish writing the last entry
+            zip.closeEntry();
+
+            // Before the CEN headers are written, set the extra data on each entry
+            byte[] extra = makeLargeExtraField();
+            for (ZipEntry entry : entries) {
+                entry.setExtra(extra);
+            }
+        }
+    }
+
+    /**
+     * Validates that a ZipException is thrown with the expected message when
+     * the ZipFile is initialized with a ZIP whose CEN exeeds {@link #MAX_CEN_SIZE}
+     */
+    @Test
+    public void centralDirectoryTooLargeToFitInByteArray() {
+        ZipException ex = expectThrows(ZipException.class, () -> new ZipFile(hugeZipFile));
+        assertEquals(ex.getMessage(), "invalid END header (central directory size too large)");
```

Comment:
Could be have the expected message a class constant so that if this message ever changes it is easier to find/update

Reply:
Extracted the expected ZipException message to a class constant.